components/bl-granim-component/src/index.js

```diff
@@ -0,0 +1,132 @@
+import { useEffect, useMemo, useRef } from 'react';
+
+import { s } from './lib/granim.min';
+
+const { cn } = BackendlessUI.CSSUtils;
+const DIRECTIONS = ['left-right', 'diagonal', 'top-bottom', 'radial',  'custom'];
+const DIRECTIONS_MAP = mapFromArray(DIRECTIONS);
+const BLENDING_MODES = [
+  'multiply', 'screen', 'normal', 'overlay', 'darken', 'lighten', 'lighter', 'color-dodge', 'color-burn',
+  'hard-light', 'soft-light', 'difference', 'exclusion', 'hue', 'saturation', 'color', 'luminosity'];
+const BLENDING_MODES_MAP = mapFromArray(BLENDING_MODES);
+const DEFAULT_CUSTOM_DIRECTION = {
+  x0: '0px',
+  y0: '0px',
+  x1: '100%',
+  y1: '100%'
+};
+const DEFAULT_STATE = {
+  gradients: [
+    ['#834d9b', '#d04ed6'],
+    ['#1CD8D2', '#93EDC7']],
+  transitionSpeed: 5000,
+  loop: true
+};
+
+export default function GranimComponent({ component, elRef, eventHandlers, pods }) {
+  const {
+    style, classList, display, direction, customDirection, states, isPausedWhenNotInView, scrollDebounceThreshold,
+    stateTransitionSpeed, imageSource, imagePosX, imagePosY, stretchModeX, stretchModeY, blendingMode,
+  } = component;
+  const { onStart, onGradientChange, onEnd } = eventHandlers;
+
+  const image = useMemo(() => {
+    return (imageSource && {
+      source: imageSource,
+      position: [imagePosX, imagePosY],
+      stretchMode: [stretchModeX, stretchModeY],
+      blendingMode,
+    });
+  }, [imageSource, imagePosX, imagePosY, stretchModeX, stretchModeY, blendingMode]);
+
+  const canvasInstRef = useRef(null);
+  const canvasRef = useRef(null);
+
+  useEffect(() => {
+    if(states && !canvasInstRef.current) {
+      if(!states['default-state']) {
+        states['default-state'] = DEFAULT_STATE;
+      }
```

Comment:
Not sure, but how about ensure states?

If there is empty states, just ensure them in additional useEffect

Reply:
Here the point is not in empty states, but in the active state, and, if the user deletes the default one and does not assign another one, this one will be applied. The states themselves can be as many as the user wishes.